### ISSUE TITLE
Clone step fails when the git host has no public DNS record

### ISSUE THREAD

**Issue by nodiscc** [woodpecker-ci/woodpecker#1171]
### Component

agent

### Describe the bug

Woodpecker pipelines triggered by a webhook from my gitea instance fail at the implicit `clone` step:

```
Initialized empty Git repository in /woodpecker/src/git.example.org/myusername/debian-live-config/.git/
+ git remote add origin https://git.example.org/myusername/debian-live-config.git
+ git fetch --no-tags origin +refs/pull/112/head:
fatal: unable to access 'https://git.example.org/myusername/debian-live-config.git/': Could not resolve host: git.example.org
exit status 128
```

The gitea instance `git.example.org` runs on the same host as woodpecker, and has no public DNS record. There is a `127.0.0.1 git.example.org` entry in the `/etc/hosts` file on the server and all clients that need to access gitea, but the container running the clone step does not use the Docker host's `/etc/hosts` file.

### System Info

```shell
{"source":"https://github.com/woodpecker-ci/woodpecker","version":"0.15.3"}
```

### Additional context

**How to reproduce:**

- Everything running on the same host
- Docker installed
- Gitea running as systemd service and listening on port 3003/tcp
- Apache reverse proxy forwarding requests for `git.example.com` to localhost:3003 [[1]](https://github.com/nodiscc/xsrv/blob/master/roles/gitea/templates/etc_apache2_sites-available_gitea.conf.j2)
- Apache reverse proxy forwarding requests for `ci.example.com` to localhost:8721 [[2]](https://github.com/nodiscc/xsrv/blob/master/roles/gitea/templates/etc_apache2_sites-available_gitea.conf.j2)
- self-signed certificates generated for `git.example.com` and `ci.example.com`

My setup is managed by [this ansible role](https://github.com/nodiscc/xsrv/tree/woodpecker/roles/woodpecker) but is easily reproducible by hand:

```bash
# create users
$ sudo adduser --home /var/lib/woodpecker-server --shell /usr/sbin/nologin woodpecker-server
$ sudo adduser --home /var/lib/woodpecker-agent --shell /usr/sbin/nologin woodpecker-agent
$ sudo usermod -G docker woodpecker-agent

# download and install woodpecker-server from binary
$ wget https://github.com/woodpecker-ci/woodpecker/releases/download/v0.15.3/woodpecker-server_linux_amd64.tar.gz
$ tar -zxvf woodpecker-server_linux_amd64.tar.gz
$ sudo cp woodpecker-server /usr/local/bin/woodpecker-server

# download and install woodpecker-agent from binary
$ wget https://github.com/woodpecker-ci/woodpecker/releases/download/v0.15.3/woodpecker-agent_linux_amd64.tar.gz
$ tar -zxvf woodpecker-agent_linux_amd64.tar.gz
$ sudo cp woodpecker-agent /usr/local/bin/woodpecker-agent
```

Create systemd services:

```ini
# $ sudo $EDITOR /etc/systemd/system/woodpecker-server.service
[Unit]
Description=woodpecker-server - continuous integration server
After=syslog.target
Requires=network-online.target
After=network-online.target

[Service]
RestartSec=2s
Type=simple
User=woodpecker-server
Group=woodpecker-server
WorkingDirectory=/var/lib/woodpecker-server
ExecStart=/usr/local/bin/woodpecker-server
EnvironmentFile=/etc/woodpecker-server.conf
Restart=on-failure
StartLimitIntervalSec=10s
StartLimitBurst=4

[Install]
WantedBy=multi-user.target
```

```ini
# $ sudo $EDITOR /etc/systemd/system/woodpecker-agent.service
[Unit]
Description=woodpecker-agent - continuous integration agent
After=syslog.target
After=network.target

[Service]
RestartSec=2s
Type=simple
User=woodpecker-agent
Group=woodpecker-agent
WorkingDirectory=/var/lib/woodpecker-agent
ExecStart=/usr/local/bin/woodpecker-agent
EnvironmentFile=/etc/woodpecker-agent.conf
Restart=on-failure
#StartLimitIntervalSec=10s
StartLimitBurst=4

[Install]
WantedBy=multi-user.target
```

Configure woodpecker:

```bash
# $ sudo $EDITOR /etc/woodpecker-server.conf
WOODPECKER_GITEA=true
WOODPECKER_GITEA_URL=https://git.example.org
WOODPECKER_GITEA_CLIENT=aaaabbbCCCddDeeefFFffGG
WOODPECKER_GITEA_SECRET=HHhhhIIiiiIJJJJjjKkkKKK
WOODPECKER_GITEA_SKIP_VERIFY=true
WOODPECKER_LOG_LEVEL=warn
WOODPECKER_SERVER_ADDR=:8721
WOODPECKER_GRPC_ADDR=:8722
WOODPECKER_ADMIN=myusername
WOODPECKER_AGENT_SECRET=LLlLLllLMmmNNnnNNOoOOOPP
WOODPECKER_HOST=https://ci.example.org

# $ sudo $EDITOR /etc/woodpecker-agent.conf
WOODPECKER_SERVER=localhost:8722
WOODPECKER_AGENT_SECRET=LLlLLllLMmmNNnnNNOoOOOPP
WOODPECKER_LOG_LEVEL=warn
WOODPECKER_HOSTNAME=home.example.org
WOODPECKER_MAX_PROCS=2
WOODPECKER_BACKEND=docker

# $ sudo $EDITOR /etc/hosts
127.0.0.1    git.example.org
127.0.0.1    ci.example.org
```

```bash
$ sudo systemctl daemon-reload
$ sudo systemctl enable --now woodpecker-server woodpecker-agent
```

- On a client with appropriate `/etc/hosts` entries, open https://ci.example.org in a browser
- Login through gitea/oauth2
- Enable a repository (the repository used in this example is [this one](https://github.com/nodiscc/debian-live-config/tree/ci-woodpecker), includes a basic [.woodpecker.yml](https://github.com/nodiscc/debian-live-config/blob/woodpecker/.woodpecker.yml))
- Commit/push anything to the repository, the automatically added webhook triggers a pipeline:

```sh
# clone
+ git init -b gitlab-ci
Initialized empty Git repository in /woodpecker/src/git.example.org/myusername/debian-live-config/.git/
+ git remote add origin https://git.example.org/myusername/debian-live-config.git
+ git fetch --no-tags origin +gitlab-ci:
fatal: unable to access 'https://git.example.org/myusername/debian-live-config.git/': Could not resolve host: git.example.org
exit status 128
```

And the pipeline fails.

Is there a way around this problem?

I could not find a way to make woodpecker run all containers with an extra option like [`--add-host git.example.org:172.18.0.1`](https://docs.docker.com/engine/reference/commandline/run/#options) (docker bridge IP), but I think this would solve the problem elegantly.

Docker has no such option either (add custom hosts entries to all containers). I can't/do not want to run a separate DNS server to workaround this issue.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A join our [Discord Chat Server](https://discord.gg/fcMQqSMXJy) or the [Matrix room](https://matrix.to/#/#woodpecker:matrix.org).

**Comment by 6543** [comment]
It use the gitea api responce to determine where to clone from

So if you would set the hostname to the internal ip that would help, but that will make responce if exported via reverse proxy invalid for the outside ...

**Comment by metzgerd** [comment]
I do have the same issue with my docker setup.

I thought I solved the dns issue while adding an alias to the nginx server:
```
  nginx:
    container_name: nginx
    image: nginx
    ports:
      - 80:80
      - 443:443
    volumes:
      - ./nginx.conf:/etc/nginx/conf.d/default.conf:ro
      - ./cert.crt:/etc/nginx/certs/cert.crt:ro
      - ./cert.key:/etc/nginx/certs/cert.key:ro
    networks:
      woodpecker-net:
        aliases:
          - git.example.com
          - ci.example.com
    depends_on:
      - gitea
      - woodpecker-server
    restart: unless-stopped
```
Inside the gitea container I can ping the alias name. Looks like there is no /bin/sh inside the woodpecker container to try the dns resolution. I still get following error:
```
+ git fetch --no-tags origin +refs/heads/main:
fatal: unable to access 'https://git.example.com/test/test.git/': Could not resolve host: git.example.com 
```
tcpdump shows that the dns request is going to the public dns server.

**Comment by 6543** [comment]
@metzgerd use the alpine version, to debug that stuff :) it has /bin/sh

**Comment by metzgerd** [comment]
Thank you. The alpine version helped to troubleshoot :-)

I can ping the nginx container from the agent and the server container and reach the nginx container via the internal IP. So the aliases setup in docker-compose works.
```
/bin # ping git.example.com
PING git.example.com (172.18.0.7): 56 data bytes
64 bytes from 172.18.0.7: seq=0 ttl=64 time=0.108 ms
```
But during clone on the pipeline I get the dns request to a public dns server.
```
6:26:22.152149 IP docker_server.59514 > one.one.one.one.domain: 15440+ AAAA? git.example.com. (35)
16:26:22.152252 IP docker_server.40932 > one.one.one.one.domain: 15248+ A? git.example.com. (35)
```
I changed the DNS settings of the VM to my internal DNS server and now it's working. Except the self signed certificate issue I mentioned in another post.

Looks like go is not using the internal docker dns server.

**Comment by 6543** [comment]
that's a agent-backend specific issue ... yes it just create a new network for each pipeline completly independent from the docker-compose env the agent runs in ... and so it only know what the host/network do provide as dns ... based on the default docker-config on the pc the agent run on ...

**Comment by 6543** [comment]
documented that feature request as https://github.com/woodpecker-ci/woodpecker/issues/1180

**Comment by nodiscc** [comment]
> adding an alias to the nginx server

My reverse proxy is an apache instance not hosted in docker, hence I cannot use this workaround.

**Comment by nodiscc** [comment]
I think the best way to solve my problem would be passing custom `--extra-hosts` (point the gitea hostname to the docker bridge IP) when the `clone` container starts. The docker backend [seems to support launching containers with extra hosts](https://github.com/woodpecker-ci/woodpecker/blob/680d003a29513f02b5e355c25672329b3de4ff97/pipeline/backend/docker/convert.go#L71) but I could not find how to pass a custom value there.

**Comment by nodiscc** [comment]
I am no longer using Woodpecker, closing.

**Comment by fredleb** [comment]
> I am no longer using Woodpecker, closing.

Since I am still facing that problem, what are you using now ? Have you found a better solution ?

**Comment by nodiscc** [comment]
> what are you using now ?

Depending on use case:
- Currently trying out [Gitea Actions](https://docs.gitea.com/next/usage/actions/overview), gitea's built-in CI/CD system with varying degrees of success (it's still in preview and planned to get out ouf preview status for the next-next release). I think I will settle with it in the long run.
- For now, my gitea repositories are mirrored (push mirror) to [gitlab.com](https://gitlab.com/users/nodiscc/projects), which in turn runs pipelines on Gitlab CI. Integration with my gitea instance is limited, in my gitea Pull Request templates I include an image/link like `[![](https://gitlab.com/nodiscc/PROJECT/badges/BRANCH/pipeline.svg)](https://gitlab.com/nodiscc/PROJECT/-/pipelines)` which lets me see the branch build status from my PR description. I used to run a personal Gitlab instance but it consumed too much resources for my simple needs. In a larger setup, I would still go with it.
- I try to keep all testing logic outside of CI-system-specific configuration files, into build scripts/Makefiles, so I can always run `make test` locally. GH actions/Gitlab CI/Gitea Actions are just a thin wrapper around that command. I experimented with post-commit/post-receive git hooks to run tests/deployments automatically after each commit, but found it better to get in the habit/discipline of running tests regularly (it's just one command).
- Over the years I tried drone, woodpecker, jenkins, concourse, laminar, buildbot... and found them all lacking in some way. So it's back to `make`, git hooks, gitlab mirrors, and progressively migrating to Gitea Actions.
